cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java

```diff
@@ -122,6 +130,41 @@ public static DataprocClient fromConf(DataprocConf conf) throws IOException, Gen
     String subnet = conf.getSubnet();
     Network networkInfo = getNetworkInfo(projectId, network, compute);
 
+    PeeringState state = peeringState(systemNetwork, systemProjectId, networkInfo, compute);
+
+    if (privateInstance) {
+      // private instance must have a peering established
+      verifyPeering(conf, network, systemNetwork, projectId, systemProjectId, state);
+      if (conf.isPreferExternalIP()) {
+        // When prefer external IP is set to true it means only Dataproc external ip can be used to for communication
+        // the instance being private instance is incapable of using external ip for communication
+        throw new IllegalArgumentException("The instance is incapable of using external ip for communication " +
+                                             "with Dataproc cluster. Please correct profile configuration by " +
+                                             "deselecting preferExternalIP.");
+      }
+    } else if (conf.isPreferExternalIP() && state == PeeringState.ACTIVE) {
+      // Peering is setup between the system network and customer network and is in ACTIVE state.
+      // However user has selected to preferred external IP the instance. This is not a private instance and is
+      // capable of communicating with Dataproc cluster with external ip so just add warning message indicating that
+      // internal IP can be used.
+      LOG.warn(String.format("VPC Peering from network '%s' in project '%s' to network '%s' " +
+                               "in project '%s' is in the ACTIVE state. Prefer External IP can be set to false " +
+                               "to launch Dataproc clusters with internal IP only.", systemNetwork,
+                             systemProjectId, network, projectId));
+    }
+
+    boolean useInternalIP;
+    if (privateInstance) {
+      // if instance is private then it is only capable of using private ip to communicate with Dataproc
+      useInternalIP = true;
+    } else {
+      // Use internal IP for the Dataproc cluster if user has not preferred external IP and
+      // (CDAP is running in the same customer project as Dataproc is going to be launched or
+      // Network peering is done between customer network and system network and is in ACTIVE mode).
+      useInternalIP = !conf.isPreferExternalIP()
+        && ((network.equals(systemNetwork) && projectId.equals(systemProjectId)) || state == PeeringState.ACTIVE);
```

Comment:
Is the systemNetwork the GCP network that CDAP itself is using? If that's the case, why it has to be the same as the Dataproc network?

Reply:
Yes this check is there for EAP and other setup where it runs in customer project.